docs/network-layer.md

```diff
@@ -0,0 +1,70 @@
+
+1. The client versions are compatible
+2. The wire versions are compatible
+3. The peers have *compatible* genesis blocks. *Compatible* in this context means that one of the other must provide a set of *genesis blocks* which should be a subset of others.
+
+When the *handshakes* has been successfully completed then the peers will *promote* each other into their corresponding lists of peers, i.e. the peers are now connected and can communicate with each other, hence the peers can now relay messages between them, carry out *consensus* and *catchup* etc.
+
+## Control Flow
+When the node is running, it continously checks for incoming connections and polls for events related to the already established connections. Furthermore the node performs several housekeeping tasks wrt. networking related matters. More specifically the node checks that its current connections are *healthy* i.e., verifying reasonable latency, removing connections of which a handshake is missing etc. If any connections are deemed *unhealthy* then the node will close the connection and remove the peer from its set of peers. Note that depending on the level of *unhealthy* the node will perhaps *soft ban* a peer (more on this below).
+
+
+If the node has exceeded its maximum amount of peers it will remove a random subset of its peers. Note that the node will never remove any of the *pinned peers*. 
+The node then proceeds to check if it should unban any of the *soft banned* peers and does so if that is the case. 
+
+Afterwards the node may try to *re-bootstrap* again (if the node did not recently *bootstrap*). Finally if the node still lacks peers then it will send out *GetPeers* messages to its peers.
+
+
+The node listens on its established connections and processes the received messages accordingly. If the message was *re-broadcastable*, then the node will relay the message to its own peers such that the message will propagate throughout the network. Before processing the message the node asserts that it has not yet seen the message before or otherwise the node will not process the *duplicated* message. 
+
+The control flow of the *network layer* can be summarized as follows:
+
+1. Listen for incoming connections
+2. Process network messages
+3. (Optionally) Carry out housekeeping tasks if a suffcient time period has passed since last time a housekeeping was done.
+4. (Optionally) Establish new outgoing connections if required
+5. Repeat
+
+## Processing Network Messages
+At the time of the connection establishment (*pre-handshake*) the connection is being registered for emitting *events*, hence a network message has been written to the underlying TCP socket. When such an event occurs, then the node will either dispatch the networking message to the *consensus layer* or immediately handle the message if it was not. And example of the former could be that the node was requested for peers by the *GetPeers* message, responding to a *Ping* etc. 
```

Comment:
I think the first sentence is confusing.  It has both too many, and too little detail to be understandable.

The listening for events part is a low-level detail so that we only read from connections that have something to read from.

I think that detail should be separated from the more high-level processing for messages which happens after a message is read. This part would be the same regardless of how the low-level socket reads happen.